runtime/src/accounts.rs

```diff
@@ -287,6 +293,40 @@ impl Accounts {
         }
     }
 
+    fn account_shared_data_from_program(
+        key: &Pubkey,
+        feature_set: &FeatureSet,
+        program: &LoadedProgram,
+        program_accounts: &HashMap<Pubkey, &Pubkey>,
+    ) -> Result<AccountSharedData> {
+        // Check for tombstone
+        if let LoadedProgramType::Invalid(reason) = &program.program {
+            match reason {
+                InvalidProgramReason::FailedToCompile => {
+                    Err(TransactionError::InstructionError(0, InvalidAccountData))
+                }
+                InvalidProgramReason::Closed => Err(TransactionError::InvalidProgramForExecution),
+                InvalidProgramReason::DelayVisibility => {
+                    debug_assert!(
+                        feature_set.is_active(&delay_visibility_of_program_deployment::id())
+                    );
+                    Err(TransactionError::InstructionError(0, InvalidAccountData))
```

Comment:
Technically this should also contain the index of the instruction it failed in. But I would also be fine with returning `TransactionError::InvalidProgramForExecution` instead.